CMSgov/easi-app
# Update GH Actions workflows to remove deprecated commands and use Node 16

# EASI-2406

Tech debt ticket to address a couple of changes by GitHub related to Actions:
    - [Deprecating `save-state` and `set-output` commands](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/). These will be fully disabled on May 31, 2023.
    - [Run actions on Node 16 instead of Node 12](https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/). Actions will be forcibly migrated to Node 16 by summer 2023.
Clay and I paired to incorporate some other improvements.

## Changes and Description
- Change our own uses of `set-output` (all used for caching `node_modules` for Yarn) to use the new `$GITHUB_OUTPUT` environment file.
- Upgrades dependencies using `set-output` and Node 12:
    - [amazon-ecr-login](https://github.com/aws-actions/amazon-ecr-login) upgraded from v1.5.1 to v1.5.3
    - [setup-go](https://github.com/actions/setup-go) upgraded from v3.3.0 to v3.4.0
    - [configure-aws-credentials](https://github.com/aws-actions/configure-aws-credentials) upgraded to the current tip of their `v1-node16` branch - they're waiting to do a major release before committing to upgrading to Node 16 only. See [this comment](https://github.com/aws-actions/configure-aws-credentials/issues/489#issuecomment-1278145876) on their repo for more info.
- All references to other actions now reference by major version number instead of by Git commit hash.
- Uses built-in caching for `setup-node` and `setup-go`, rather than writing our own caching logic. I've confirmed that both are being used, though the Go caching only helps the `server_test` and `lint` jobs, not actually building the backend container image. I think this is because that's done in our Dockerfile, which runs `go mod download` itself.

## How to test this change

* Look at the workflow run for the checks on this PR and notice the lack of warnings; compare to the warnings on a recent run such as [this one](https://github.com/CMSgov/easi-app/actions/runs/3633171806) (bottom of the page, under "Annotations") 
* Check https://github.com/CMSgov/easi-app/actions/caches and see that caches have been created, associated with this PR.

## Notes
* There's one remaining warning about `ubuntu-latest workflows` switching to `ubuntu-22.04`; I created https://github.com/CMSgov/easi-app/pull/1850, branched off this branch, to explicitly test our actions on `ubuntu-22.04`. All checks passed, it looks fine. `ubuntu-latest` will switch to using `ubuntu-22.04` automatically at some point (see [this announcement](https://github.blog/changelog/2022-08-09-github-actions-ubuntu-22-04-is-now-generally-available-on-github-hosted-runners/)), the warning will go away then and we shouldn't experience any issues.
